crates/fluvio-spu/src/kv/consumer.rs

```diff
@@ -0,0 +1,295 @@
+    }
+
+    async fn delete(&mut self, key: &Key) -> Result<()> {
+        trace!(?key, "delete");
+        let result = self.kv.delete(key).await;
+        self.changes_since_flush.add_assign(1);
+        self.maybe_flush().await?;
+        result
+    }
+
+    async fn put(&mut self, key: impl Into<Key>, value: impl Into<Consumer>) -> Result<()> {
+        let key = key.into();
+        let value = value.into();
+        trace!(?key, ?value, "put");
+        let result = self.kv.put(key, value).await;
+        self.changes_since_flush.add_assign(1);
+        self.maybe_flush().await?;
+        result
+    }
+}
+
+impl Key {
+    pub(crate) fn new(replica_id: impl Into<ReplicaKey>, consumer_id: impl Into<String>) -> Self {
+        Self {
+            replica_id: replica_id.into(),
+            consumer_id: consumer_id.into(),
+        }
+    }
+}
+impl Consumer {
+    pub(crate) fn new(offset: i64, ttl: Duration) -> Self {
+        let now = now_timestamp();
+        Self::with(offset, ttl, now + ttl.as_secs())
+    }
+
+    pub(crate) fn with(offset: i64, ttl: Duration, expire_time: TimestampSecs) -> Self {
+        Self {
+            offset,
+            ttl,
+            expire_time,
+        }
+    }
+}
+
+impl From<ConsumerStorage> for SharableConsumerStorage {
+    fn from(value: ConsumerStorage) -> Self {
+        Self(Arc::new(RwLock::new(value)))
+    }
+}
+
+impl SharableConsumerStorage {
+    pub async fn get_by(
+        &self,
+        topic: String,
+        partition: PartitionId,
+        consumer_id: String,
+    ) -> Result<Option<Consumer>> {
+        let replica: ReplicaKey = (topic, partition).into();
+        let key = Key::new(replica, consumer_id);
+        self.get(&key).await
+    }
+
+    pub async fn get(&self, key: &Key) -> Result<Option<Consumer>> {
+        self.0.read().await.get(key).await
+    }
+
+    pub async fn delete(&self, key: &Key) -> Result<()> {
+        self.0.write().await.delete(key).await
+    }
+
+    pub async fn put(&self, key: impl Into<Key>, value: impl Into<Consumer>) -> Result<()> {
+        self.0.write().await.put(key, value).await
+    }
+}
+
+fn now_timestamp() -> TimestampSecs {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .unwrap_or_default()
+        .as_secs()
+}
+
+fn is_not_expired(record: &Consumer) -> bool {
+    let now = now_timestamp();
+    now <= record.expire_time
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{env::temp_dir, path::Path};
+
+    use fluvio_controlplane::replica::Replica;
+    use fluvio_storage::config::ReplicaConfig;
+    use fluvio_types::PartitionId;
+    use flv_util::fixture::ensure_clean_dir;
+
+    use crate::{
+        config::ReplicationConfig, storage::SharableReplicaStorage,
+        control_plane::StatusMessageSink,
+    };
+
+    use super::*;
+
+    #[fluvio_future::test]
+    async fn test_expiration_checked() {
+        //given
+        let leader = create_replica("test_expiration_checked").await;
```

Comment:
change name to something like `offset_replica`

Reply:
Renamed